cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueDebugger.java

```diff
@@ -0,0 +1,308 @@
+
+    byte[] stateColumnName = Bytes.add(QueueEntryRow.STATE_COLUMN_PREFIX,
+                                       Bytes.toBytes(consumerConfig.getGroupId()));
+
+
+    int distributorBuckets = queueClientFactory.getDistributorBuckets(hTable.getTableDescriptor());
+    ShardedHBaseQueueStrategy queueStrategy = new ShardedHBaseQueueStrategy(distributorBuckets);
+
+    Scan scan = new Scan();
+    scan.setStartRow(start.getStartRow());
+    if (end != null) {
+      scan.setStopRow(end.getStartRow());
+    }
+    scan.addColumn(QueueEntryRow.COLUMN_FAMILY, QueueEntryRow.DATA_COLUMN);
+    scan.addColumn(QueueEntryRow.COLUMN_FAMILY, QueueEntryRow.META_COLUMN);
+    scan.addColumn(QueueEntryRow.COLUMN_FAMILY, stateColumnName);
+    scan.setCaching(100);
+    scan.setMaxVersions(1);
+
+    LOG.info("Scanning section with scan: {}", GSON.toJson(scan));
+    QueueStatistics stats = new QueueStatistics();
+
+    ConsumerConfig consConfig = new ConsumerConfig(consumerConfig, 0);
+    ResultScanner scanner = queueStrategy.createHBaseScanner(consConfig, hTable, scan, 100, true);
+
+    Function<byte[], byte[]> rowKeyConverter = ShardedHBaseQueueStrategy.ROW_KEY_CONVERTER;
+
+    Result result;
+    while ((result = scanner.next()) != null) {
+      NavigableMap<byte[], byte[]> columns = result.getFamilyMap(QueueEntryRow.COLUMN_FAMILY);
+      byte[] rowKey = rowKeyConverter.apply(result.getRow());
+      visitRow(stats, tx, rowKey, columns.get(stateColumnName), queueRowPrefix.length);
+    }
+    return stats;
+  }
+
+  /**
+   * @param tx the transaction
+   * @param rowKey the key of the row
+   * @param stateValue the value of the state column in the row
+   * @param queueRowPrefixLength length of the queueRowPrefix
+   */
+  private void visitRow(QueueStatistics stats, Transaction tx, byte[] rowKey,
+                        byte[] stateValue, int queueRowPrefixLength) {
+    if (stateValue == null) {
+      stats.countUnprocessed();
+      return;
+    }
+
+    ConsumerEntryState state = QueueEntryRow.getState(stateValue);
+    if (state == ConsumerEntryState.PROCESSED) {
+      long writePointer = QueueEntryRow.getWritePointer(rowKey, queueRowPrefixLength);
+      if (tx.isVisible(writePointer)) {
+        stats.countProcessedAndVisible();
+      } else {
+        stats.countProcessedAndNotVisible();
+      }
+    }
+  }
+
+  /**
+   *
+   */
+  private static final class QueueStatistics {
+
+    private long unprocessed;
+    private long processedAndVisible;
+    private long processedAndNotVisible;
+
+    private QueueStatistics() {
+    }
+
+    private QueueStatistics(long unprocessed, long processedAndVisible, long processedAndNotVisible) {
+      this.unprocessed = unprocessed;
+      this.processedAndVisible = processedAndVisible;
+      this.processedAndNotVisible = processedAndNotVisible;
+    }
+
+    public static QueueStatistics of(QueueStatistics...stats) {
```

Comment:
Instead of having this method, why not have the `scanQueue` method takes a `QueueStatistics` object as argument and modify it instead of returning a new one?